### PR TITLE
mrd-625 and mrd-630 DNTR document component along with `why recall considered` information.

### PR DESCRIPTION
**Pull request by bills7281**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/controller/RecommendationController.kt [hunk 1]
```diff
@@ -82,13 +82,29 @@ internal class RecommendationController(
...
     } catch (e: UserAccessException) {
-      ResponseEntity(PartAResponse(Gson().fromJson(e.message, UserAccessResponse::class.java)), FORBIDDEN)
+      ResponseEntity(DocumentResponse(Gson().fromJson(e.message, UserAccessResponse::class.java)), FORBIDDEN)
+    }
+    return responseEntity
+  }
+
+  @PreAuthorize("hasRole('ROLE_MAKE_RECALL_DECISION')")
+  @PostMapping("/recommendations/{recommendationId}/task-list-no-recall")
+  @Operation(summary = "Generates a Dntr document")
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/controller/RecommendationController.kt:98]: Guess the `Dntr` should be uppercase here

Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/controller/RecommendationController.kt]: ok, fixed!! :)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import uk.gov.justice.digital.hmpps.makerecalldecisionapi.domain.makerecalldecisions.recommendation.DntrData
+import uk.gov.justice.digital.hmpps.makerecalldecisionapi.jpa.entity.RecommendationEntity
+import uk.gov.justice.digital.hmpps.makerecalldecisionapi.util.MrdTextConstants.Constants.EMPTY_STRING
+import uk.gov.justice.digital.hmpps.makerecalldecisionapi.util.RecommendationToDocumentMapper
+
+@Service
+internal class DntrTemplateReplacementService {
+  val resource = ClassPathResource("DNTR Template.docx")
+
+  fun generateDocFromTemplate(recommendation: RecommendationEntity): String {
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt:15]: This looks near identical to the one found in `PartATemplateReplacementService` - can we make a generic class/interface that sits over these ones?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt [hunk 3]
```diff
@@ -127,14 +128,15 @@ internal class RecommendationService(
...
+        mainAddressWherePersonCanBeFound = recommendationEntity.data.mainAddressWherePersonCanBeFound,
+        whyConsideredRecall = recommendationEntity.data.whyConsideredRecall
       )
     }
   }
 
   @OptIn(ExperimentalStdlibApi::class)
   @Transactional
-  fun updateRecommendation(jsonRequest: JsonNode?, recommendationId: Long, username: String?, userEmail: String?, isPartADownloaded: Boolean): RecommendationEntity {
+  fun updateRecommendation(jsonRequest: JsonNode?, recommendationId: Long, username: String?, userEmail: String?, isPartADownloaded: Boolean, isDNTRDownloaded: Boolean = false): RecommendationEntity {
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt:139]: Could we change to `isDntrDownloaded`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt [hunk 4]
```diff
@@ -208,6 +228,15 @@ internal class RecommendationService(
     return "NAT_Recall_Part_A_${nowDate()}_${surname}_${firstName}_$crn.docx"
   }
 
+  private fun generateDNTRFileName(recommendation: RecommendationModel): String {
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt:231]: `generateDntrFileName`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt (hunk 2 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt:15]: I had a go, but found myself forking with if statements on the output filenames and struggled generifying the mappings. I could use the strategy pattern encapsulate them both into a single dependency for the RecommendationService to use. But, that doesn't really reduce the code footprint, and I feel it's overkill as we only have 2 strategies (document types).

I am totally open to suggestion on this point :)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt (hunk 4 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt [hunk 5]
```diff
@@ -182,15 +187,30 @@ internal class RecommendationService(
   }
 
   @OptIn(ExperimentalStdlibApi::class)
-  fun generatePartA(recommendationId: Long, username: String?, userEmail: String?): PartAResponse {
+  fun generateDNTR(recommendationId: Long, username: String?): DocumentResponse {
```
[the remark this answers is not in the document]
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt (hunk 3 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/RecommendationService.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt (hunk 2 above)
Comment (bills7281) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt:15]: I had a go, but found myself needing to pass in the filename/data which blurred the single responsibility of the RecommendationService. I also struggled with generifying the mappings. 

I could use the strategy pattern encapsulate them both into a single dependency for the RecommendationService to use. But, that doesn't really reduce the code footprint, and I feel it's overkill as we only have 2 strategies (document types).

I am totally open to suggestion on this point :)

Comment (jackmaloney) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/service/DntrTemplateReplacementService.kt:15]: Ok maybe let's get it merged and we can refactor it at some point

**Review by jackmaloney — APPROVED** [review]

File: src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationForNoRecallRequest.kt [hunk 6]
```diff
@@ -0,0 +1,233 @@
+package uk.gov.justice.digital.hmpps.makerecalldecisionapi.integration.requests.makerecalldecisions
+
+import uk.gov.justice.digital.hmpps.makerecalldecisionapi.jpa.entity.Status
+
+fun updateRecommendationForNoRecallRequest(status: Status = Status.DRAFT) = """
```
Comment (jackmaloney) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/integration/requests/makerecalldecisions/UpdateRecommendationForNoRecallRequest.kt:5]: I think there is some data in here that isn't relevant to the 'No recall' journey, is it worth updating to keep it clean?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/controller/RecommendationController.kt [hunk 7]
```diff
@@ -82,13 +82,29 @@ internal class RecommendationController(
...
       ResponseEntity(recommendationService.generatePartA(recommendationId, authenticationFacade.currentNameOfUser, createPartARequest.userEmail), OK)
     } catch (e: UserAccessException) {
-      ResponseEntity(PartAResponse(Gson().fromJson(e.message, UserAccessResponse::class.java)), FORBIDDEN)
+      ResponseEntity(DocumentResponse(Gson().fromJson(e.message, UserAccessResponse::class.java)), FORBIDDEN)
+    }
+    return responseEntity
+  }
+
+  @PreAuthorize("hasRole('ROLE_MAKE_RECALL_DECISION')")
+  @PostMapping("/recommendations/{recommendationId}/task-list-no-recall")
```
Comment (jackmaloney) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/makerecalldecisionapi/controller/RecommendationController.kt:97]: Just wondering if the text 'task-list' is right here? We don't have it in the Part A endpoint name so thinking we should make it consistent? Maybe add 'task-list' to the Part A endpoint or remove from his one?

**Review by jackmaloney — APPROVED** [review]

**Pull request merged**